### PR TITLE
Fix price calculation for mixed-tier winners

### PR DESCRIPTION
**Pull request by jimj**
Previously if there were more than 1 item, the top winner was compared
against the 2nd winner to determine their item price.

This change instead always compares the winner against the first loser's
price for price calculation

### CODE REVIEW

File: DiscordDkpBot/Commands/CancelBidCommand.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@ namespace DiscordDkpBot.Commands
 	[UsedImplicitly]
 	public class CancelBidCommand : IDmCommand
 	{
-		private static readonly Regex pattern = new Regex(@"^""(?<Item>.+)""\s+cancel\s*$", RegexOptions.IgnoreCase);
+		private static readonly Regex pattern = new Regex(@"^\s*\S(?<Item>.+)\S\s+cancel\s*$", RegexOptions.IgnoreCase);
```
Comment (SterlingNerd) [inline DiscordDkpBot/Commands/CancelBidCommand.cs:18]: If we're going to do this. we should do it to ALL the commands.

File: DiscordDkpBot/Auctions/AuctionProcessor.cs [hunk 2]
```diff
@@ -188,27 +188,15 @@ private List<WinningBid> CalculatePrices (ICollection<AuctionBid> winners, IColl
...
-						loser = losers.OrderBy(x=>x).FirstOrDefault();
-
-						//If the loser tied a winner, that adjusts the final price.
-						//But only if they didn't outbid their cap'
-						if (loser?.BidAmount == winners.LastOrDefault()?.BidAmount
-						&& loser.BidAmount <= loser.MaxBid)
-						{
-							loserTied = true;
-						}
+						loserTied = true;
```
Comment (SterlingNerd) [inline DiscordDkpBot/Auctions/AuctionProcessor.cs:199]: This breaks existing loot rules. Specifically if there's 3 bids on on two items. 
See AuctionProcessorTests.CalculateWinners_TwoItems_ThreeBids.

What is the goal of these changes. You should build a test case in AuctionProcessorTests to represent it, and ensure your updates don't break other scenarios. If there's a conflict, we might need to consider refactoring to support different rulesets.

File: DiscordDkpBot/Commands/DkpCheckCommand.cs [hunk 3]
```diff
@@ -70,22 +70,12 @@ public async Task<bool> TryInvokeAsync(IMessage message)
...
-				using (MemoryStream stream = new MemoryStream(Convert.FromBase64String(Resources.dkpkalmarea)))
-				{
-					await message.Channel.SendFileAsync(stream, "dkpkalmarea.gif", dkpMessage);
-				}
+			string dkpMessage = $"{character.UppercaseFirst()} has **{dkp.PointsCurrentWithTwink}** available to spend.\n```\nLifetime DKP for {character}: Earned {dkp.PointsEarnedWithTwink} - Spent {dkp.PointsSpentWithTwink} - Adjustments {dkp.PointsAdjustmentWithTwink}.```";
+			string memePath = config.MemeDkpPath + character + ".jpg"; ;
+
+			if (config.EnableMemeDkp && File.Exists(memePath)) {
+				log.LogInformation("dkp memes!");
+				await message.Channel.SendFileAsync(memePath, dkpMessage);
```
Comment (SterlingNerd) [inline DiscordDkpBot/Commands/DkpCheckCommand.cs:78]: I like this. I'm incorporating it into the next release right now.

File: DiscordDkpBotTests/Auctions/AuctionBidTests.cs [hunk 4]
```diff
@@ -105,6 +105,36 @@ public void CompareToMultiples ()
...
+
+			//Assert
+			Assert.AreEqual(mainBid, list[0]);
+			Assert.AreEqual(boxBid1, list[1]);
+			Assert.AreEqual(boxBid2, list[2]);
+			Assert.AreEqual(mainBid2, list[3]);
+			Assert.AreEqual(boxBid3, list[4]);
+
+		}	
+
```
Comment (SterlingNerd) [inline DiscordDkpBotTests/Auctions/AuctionBidTests.cs:137]: This doesn't actually add anything new. These tests are just to ensure bids get sorted properly. AuctionProcessor is where we run winner logic.

**Review by SterlingNerd — COMMENTED** [review]
In general, pull requests should contain a single change. It makes it easier to review and incorporate.

Meme improvements 
- I like these, and have separately applied them to master already

Regex change 
- (Consistency is important, if we want to change " to \S we should do it to ALL commands, not just cancel. 
- If the idea is to make it easier to cancel an auction, We might be able to wire up a reaction trigger. Things to think about.

A version change
- Generally we only do this when pushing a release. Not individual changes. (aka don't include version #s in your pull requests, let the person putting the release package together bump the numbers)

An auction logic fix.
- This is the actual goal of the pull request based on the title. The rest of the stuff was just trying to sneak in. The bug is legit and the unit test looks good. 
- But unfortunately your fix broke a separate scenario. I'm working on solving it right now.

In summary, good ideas here. But because of the issues I mentioned instead of merging this I'll simply copy/paste out of here instead.

Thanks for the help!

**Comment by SterlingNerd** [comment]
This was actually an issue I had already considered, but there was a bug in the test! The horror!

Anyway, it's fixed in f318b1e

**Comment by jimj** [comment]
> In general, pull requests should contain a single change. It makes it easier to review and incorporate.
> 
> Meme improvements
> 
>     * I like these, and have separately applied them to master already
> 
> 
> Regex change
> 
>     * (Consistency is important, if we want to change " to \S we should do it to ALL commands, not just cancel.
> 
>     * If the idea is to make it easier to cancel an auction, We might be able to wire up a reaction trigger. Things to think about.
> 
> 
> A version change
> 
>     * Generally we only do this when pushing a release. Not individual changes. (aka don't include version #s in your pull requests, let the person putting the release package together bump the numbers)
> 
> 
> An auction logic fix.
> 
>     * This is the actual goal of the pull request based on the title. The rest of the stuff was just trying to sneak in. The bug is legit and the unit test looks good.
> 
>     * But unfortunately your fix broke a separate scenario. I'm working on solving it right now.
> 
> 
> In summary, good ideas here. But because of the issues I mentioned instead of merging this I'll simply copy/paste out of here instead.
> 
> Thanks for the help!

First time I've done a pull request, which was made with the first commit only.  I didn't intend for any others to show up here, and they were made to my branch after this PR was open.

**Comment by SterlingNerd** [comment]
Makes sense. Best practices says to do each change in it's own branch partially to avoid exactly this.

File: DiscordDkpBot/Auctions/AuctionProcessor.cs (hunk 2 above)
Comment (jimj) [reply DiscordDkpBot/Auctions/AuctionProcessor.cs:199]: I find this a dubious claim as I both didn't break the existing test and also caused the new one to pass.  You don't have to take this behavior change, but we for sure want it to behave this way when there are mixed tier winners.  We manually adjust auctions before this fix.

File: DiscordDkpBot/Commands/DkpCheckCommand.cs (hunk 3 above)
Comment (jimj) [reply DiscordDkpBot/Commands/DkpCheckCommand.cs:78]: It's a lot of fun to swap memes in real time

File: DiscordDkpBotTests/Auctions/AuctionBidTests.cs (hunk 4 above)
Comment (jimj) [reply DiscordDkpBotTests/Auctions/AuctionBidTests.cs:137]: Yep, was trying to find where the bug was

File: DiscordDkpBot/Commands/CancelBidCommand.cs (hunk 1 above)
Comment (jimj) [reply DiscordDkpBot/Commands/CancelBidCommand.cs:18]: Didn't see any others with this issue.  I copied this regex from the bid creation command so smart quotes were already dealt with there.

Just a heads up I don't actually care which of these you pull or not.  We're running my fork now.

File: DiscordDkpBot/Auctions/AuctionProcessor.cs (hunk 2 above)
Comment (SterlingNerd) [reply DiscordDkpBot/Auctions/AuctionProcessor.cs:199]: Maybe I applied it wrong. I want to say there was a specific reason that extra piece of code was in there for. Maybe I forgot to add a unit test for it. In any case f318b1e should have fixed the issue we were looking at just as well.
